File: Python/needleman_wunsch.py
#!/usr/bin/env python3

import argparse
import numpy as np
import matplotlib.pyplot as plt

def check_match(a, b, match_score, mismatch_score):
    if (a == b):
        return match_score
    else:
        return mismatch_score

def generate_matrix(seq1, seq2, match_score, mismatch_score, gap_penalty):
    sim_matrix = np.zeros((len(seq2) + 1, len(seq1) + 1), dtype=int)
    for i in range(len(seq2) + 1):
        sim_matrix[i, 0] = gap_penalty * i
    for j in range(len(seq1) + 1):
        sim_matrix[0, j] = gap_penalty * j
    for seq1_col in range(len(seq1)):
        for seq2_row in range(len(seq2)):
            match = sim_matrix[seq2_row, seq1_col] + check_match(seq2[seq2_row], seq1[seq1_col], match_score, mismatch_score)
            delete = sim_matrix[seq2_row, seq1_col + 1] + gap_penalty
            insert = sim_matrix[seq2_row + 1, seq1_col] + gap_penalty
            sim_matrix[seq2_row + 1, seq1_col + 1] = max(match, delete, insert)
    return sim_matrix, sim_matrix[-1,-1]

def pairwise_align(seq1, seq2, match_score, mismatch_score, gap_penalty):
    sim_matrix,score = generate_matrix(seq1, seq2, match_score, mismatch_score, gap_penalty)
    AlignmentA = ''
    AlignmentB = ''
    col_seq1 = len(seq1)
    row_seq2 = len(seq2)

    while (col_seq1 > 0 or row_seq2 > 0):
        if (col_seq1 > 0 and row_seq2 > 0):
            diag = sim_matrix[row_seq2, col_seq1] == sim_matrix[row_seq2 - 1, col_seq1 - 1] + check_match(seq2[row_seq2 - 1], seq1[col_seq1 - 1], match_score, mismatch_score)
            delete = sim_matrix[row_seq2, col_seq1] == sim_matrix[row_seq2 - 1, col_seq1] + gap_penalty
            insert = sim_matrix[row_seq2, col_seq1] == sim_matrix[row_seq2, col_seq1 - 1] + gap_penalty
            if (insert):
                AlignmentA += seq1[col_seq1 - 1]
                AlignmentB += '-'
                col_seq1 -= 1
            elif (delete):
                AlignmentA += '-'
                AlignmentB += seq2[row_seq2 - 1]
                row_seq2 -= 1
            elif (diag):
                AlignmentA += seq1[col_seq1 - 1]
                AlignmentB += seq2[row_seq2 - 1]
                col_seq1 -= 1
                row_seq2 -= 1
            else:
                raise Exception('Stuck in an infinite loop')
        elif (col_seq1 > 0):
            AlignmentA += seq1[col_seq1 - 1]
            AlignmentB += '-'
            col_seq1 -= 1
        elif (row_seq2 > 0):
            AlignmentA += '-'
            AlignmentB += seq2[row_seq2 - 1]
            row_seq2 -= 1
        else:
            raise Exception('Stuck in an infinite loop')
    AlignmentA = AlignmentA[::-1]
    AlignmentB = AlignmentB[::-1]

    return AlignmentA, AlignmentB, score


parser = argparse.ArgumentParser()

parser.add_argument(
    '-s',
    '--sequences',
    help = 'Enter the sequences to be compared',
    nargs = 2
)

parser.add_argument(
    '-match',
    '--match_score',
    help = 'Enter the Score given on a match, default is +1',
    default = 1
)

parser.add_argument(
    '-mismatch',
    '--mismatch_score',
    help = 'Enter the Score given on a mismatch, default is -1',
    default = -1
)

parser.add_argument(
    '-g',
    '--gap_penalty',
    help = 'Enter the score given on an insertion or deletion, default is -2',
    default = -2
)

args = parser.parse_args()

if args.sequences:
    seq1 = args.sequences[0]
    seq2 = args.sequences[1]
    match_score = int(args.match_score)
    mismatch_score = int(args.mismatch_score)
    gap_penalty = int(args.gap_penalty)

    align_seq1, align_seq2, score = pairwise_align(seq1, seq2, match_score, mismatch_score, gap_penalty)

    print('----------Sequences---------------')
    print()
    print('Sequence 1: ', seq1)
    print('Sequence 2: ', seq2)
    print()
    print('----------Alignment---------------')
    print()
    print('Alignment 1: ', align_seq1)
    print('Alignment 2: ', align_seq2)
    print()
    print('Score:', score)

File: R/needleman_wunsch.r
rm(list = ls())
# Install these libraries in case they don't exist
library("stringr")
library("stringi")

# Put Sequences here
seq1 <- "ATTTCTAGGGAAGTTTCAGATGGAGAAATGTGTGGACCATTCAGGCCCTAAAAGTTGGTATAAAGAT"
seq2 <- "AAGCACCGCGTAGGCCAGCTGGCCGGATCCCGCCGTCTGTCATGGCGGCCCCCATCCTGAAAGG"

# The respective match score, mismatch score and gap penalty
match_score <- 1
mismatch_score <- -1
gap_penalty <- -2

# Breaking the string into a vector of characters - like ['A','C','T','G']
s1_v <- strsplit(seq1, NULL)[[1]]
s2_v <- strsplit(seq2, NULL)[[1]]

# Function to check whether two base pairs match or don't, and returns the respective score
check_match <- function(a, b, match, mismatch) {
  if (str_trim(a) == str_trim(b)) {
    return(match)
  } else {
    return(mismatch)
  }
}

# Function to generate the similarity matrix based on gap penalty and match/mismatch score
generate_matrix <- function(seq1, seq2, match_score, mismatch_score, gap_score) {

  sim_matrix <- matrix(0, nrow = length(seq2) + 1, ncol = length(seq1) + 1)

  for (i in 1 : ncol(sim_matrix)) {
    sim_matrix[1, i] <- gap_score * (i - 1)
  }
  for (i in 1: nrow(sim_matrix)) {
    sim_matrix[i, 1] <- gap_score * (i - 1)
  }
  for (seq1_col in 1:length(seq1)) {
    for (seq2_row in 1:length(seq2)) {
      match <- sim_matrix[seq2_row, seq1_col] +
        check_match(seq2[seq2_row], seq1[seq1_col],
                    match_score, mismatch_score)
      delete <- sim_matrix[seq2_row, seq1_col + 1] + gap_score
      insert <- sim_matrix[seq2_row + 1, seq1_col] + gap_score
      sim_matrix[seq2_row + 1, seq1_col + 1] <- max(match, delete, insert)
      rm(match, insert, delete)
    }
  }
  return(list(sim_matrix = sim_matrix,
              score = sim_matrix[nrow(sim_matrix), ncol(sim_matrix)]))
}

# Function to align the two sequences tracing back the similarity matrix
pairwise_align <- function(seq1, seq2, match_score, mismatch_score, gap_penalty) {

  result <- generate_matrix(seq1, seq2, match_score, mismatch_score, gap_penalty)
  sim_matrix <- result$sim_matrix
  alignment_a <- ""
  alignment_b <- ""
  col_seq1 <- length(seq1) + 1
  row_seq2 <- length(seq2) + 1

  while (col_seq1 > 1 || row_seq2 > 1) {
    if (col_seq1 > 1 && row_seq2 > 1) {
      diag <- sim_matrix[row_seq2, col_seq1] == sim_matrix[row_seq2 - 1,col_seq1 - 1] +
              check_match(seq2[row_seq2 - 1], seq1[col_seq1 - 1], match_score, mismatch_score)
      delete <- sim_matrix[row_seq2, col_seq1] == sim_matrix[row_seq2 - 1, col_seq1] + gap_penalty
      insert = sim_matrix[row_seq2, col_seq1] == sim_matrix[row_seq2, col_seq1 - 1] + gap_penalty

      if (insert) {
        alignment_a <- paste0(alignment_a, seq1[col_seq1 - 1])
        alignment_b <- paste0(alignment_b, "-")
        col_seq1 <- col_seq1 - 1
      } else if (delete) {
        alignment_a <- paste0(alignment_a, "-")
        alignment_b <- paste0(alignment_b, seq2[row_seq2 - 1])
        row_seq2 <- row_seq2 - 1
      } else if (diag) {
        alignment_a <- paste0(alignment_a, seq1[col_seq1 - 1])
        alignment_b <- paste0(alignment_b, seq2[row_seq2 - 1])
        col_seq1 <- col_seq1 - 1
        row_seq2 <- row_seq2 - 1
      } else {
        return()
      }
    } else if (col_seq1 > 0) {
      alignment_a <- paste0(alignment_a, seq1[col_seq1 - 1])
      alignment_b <- paste0(alignment_b, "-")
      col_seq1 <- col_seq1 - 1
    } else if (row_seq2 > 0) {
      alignment_a <- paste0(alignment_a, "-")
      alignment_b <- paste0(alignment_b, seq2[row_seq2 - 1])
      row_seq2 <- row_seq2 - 1
    } else {
      return()
    }
  }

  alignment_a <- stri_reverse(alignment_a)
  alignment_b <- stri_reverse(alignment_b)


  return(list(alignment_a = alignment_a, alignment_b = alignment_b, score = result$score))
}

# Printing the results
align <- pairwise_align(s1_v, s2_v, match_score, mismatch_score, gap_penalty)

cat("Alignment 1: ", align$alignment_a, "\n")
cat("Alignment 2: ", align$alignment_b, "\n")
cat("Score:", align$score, "\n")